src/validator.rs

```diff
@@ -188,39 +246,50 @@ impl<T> InquireLength for &[T] {
 
 /// Built-in validator that checks whether the answer is not empty.
 ///
-/// # Arguments
-///
-/// * `$message` - optional - Error message returned by the validator.
-///   Defaults to "A response is required."
-///
 /// # Examples
 ///
 /// ```
-/// use inquire::{required, validator::{StringValidator, Validation}};
+/// use inquire::validator::{StringValidator, Validation, ValueRequiredValidator};
 ///
-/// let validator: StringValidator = required!();
-/// assert_eq!(Validation::Valid, validator("Generic input")?);
-/// assert_eq!(Validation::Invalid("A response is required.".into()), validator("")?);
+/// let validator = ValueRequiredValidator::new();
+/// assert_eq!(Validation::Valid, validator.validate("Generic input")?);
+/// assert_eq!(Validation::Invalid("A response is required.".into()), validator.validate("")?);
 ///
-/// let validator: StringValidator = required!("No empty!");
-/// assert_eq!(Validation::Valid, validator("Generic input")?);
-/// assert_eq!(Validation::Invalid("No empty!".into()), validator("")?);
+/// let validator = ValueRequiredValidator::new().with_message("No empty!");
+/// assert_eq!(Validation::Valid, validator.validate("Generic input")?);
+/// assert_eq!(Validation::Invalid("No empty!".into()), validator.validate("")?);
 /// # Ok::<(), inquire::error::CustomUserError>(())
 /// ```
-#[macro_export]
-#[cfg(feature = "builtin_validators")]
-macro_rules! required {
-    () => {
-        $crate::required! {"A response is required."}
-    };
+#[derive(Clone, Default)]
+pub struct ValueRequiredValidator {
+    message: Option<String>,
+}
```

Comment:
That's an excellent idea, and good usage of `Default`/`new`. Also, can get rid of the `Option`, as we technically always have a message.

What do you think about using `Cow<'static, str>` instead of `String`? Would help to avoid allocations in many cases. On the other hand, it might be a  bit of an over-optimizations, as a terminal crate usually doesn't have to worry too much about a few allocations here and there 🤔.

For example:
```rust
use std::borrow::Cow;

struct ValueRequiredValidator {
    message: Cow<'static, str>,
}

impl ValueRequiredValidator {
    pub fn new(message: impl Into<Cow<'static, str>>) -> Self {
        Self {
            message: message.into(),
        }
    }
}

fn main() {
    // This one allocates a string and takes it as before.
    let v = ValueRequiredValidator::new(format!("hello, {}!", "World"));
    assert!(matches!(v.message, Cow::Owned(_)));

    // This one avoids an allocation and takes a reference to the static string.
    // It would be the same case for ValueRequiredValidator::default().
    let v = ValueRequiredValidator::new("hello, World!");
    assert!(matches!(v.message, Cow::Borrowed(_)));
}
```

Reply:
Leaving the memory optimization aside, I see a lot of value in making this syntax possible:

```
let v = ValueRequiredValidator::new("hello, World!");
```

_Which I just realized also was possible when you used `impl Into<String>`_

What I'm thinking about is that `'static` lifetime and a little confusion that might come with it in certain cases:

```
fn prompt(validator_message: &str) {
    // `validator_message` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
    let v = ValueRequiredValidator::new(validator_message);
    assert!(matches!(v.message, Cow::Borrowed(_)));
}
```

So we have 3 options I guess for the new argument

- `String`:
  - Pro: Straightforward, newcomers specially will never be confused which is one of my goals in making this crate focused on a  good API.
  - Con: Will always require the string to be allocated.
- `Into<String>`: I guess same as above + allowing the use of raw strings. Downside is cloning even strings already owned.
- `Into<Cow<'static, String>>` which looks good regarding the optimizations but has the potential of making the API a little bit more uninviting, specially for newcomers. There's also the fact that down the line this error message will be converted to an owned string either way on `ErrorMessage`, a public struct where I think using `Cow` would bring more complexity to the API.

In the end I'm fine with either Into<String> or String as parameters, and quite not sure if the Cow is worth it, wdyt?

Reply:
How about we stick with `Into<String>` then?

> Downside is cloning even strings already owned.
Each type automatically implements `Into<X>` for itself, and in case of `String` I believe it's a no-op. So if we pass some sort of owned string, it'll use it without an extra clone.

Reply:
Good point. I had researched before commenting but turns out I mixed Into<String> with ToString, which does make an unnecessary clone...

Let's go with Into<String>